### PR TITLE
[rust] Add fundamental types for random cut trees

### PR DESCRIPTION
**Pull request by cswiercz**
To begin merging the Rust implementation, we start with some fundamental tree building blocks.

* Nodes (Leaf, Internal)
* Bounding Boxes
* Cuts
* Point Stores 

Note that the point store uses [`slabs`](https://docs.rs/slab/0.4.2/slab/) for now.

Building:

    $ cargo build

Testing:

    $ cargo test

Documentation:

    $ cargo doc —no-deps
    $ open target/doc/rcf/index.html

> By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by cswiercz** [comment]
Minor note: some of the doc tests won’t compile until other functionality is merged.

**Comment by cswiercz** [comment]
I'm done making minor (mostly aesthetic) adjustments to the code for now. Will wait for review until making any further changes.

File: Rust/src/tree/node.rs
Comment (jotok) [inline Rust/src/tree/node.rs:202]: Could we avoid some of this repetition by having Leaf and Internal both implement a Node trait?

File: Rust/Cargo.toml
Comment (cswiercz) [reply Rust/Cargo.toml:2]: The best reference for Rust packages is [`lib.rs`](https://lib.rs/). In the eyeball norm, package names tend to be short. Related: there is a discussion [here](https://github.com/rust-lang/api-guidelines/discussions/29), referenced from the [Rust Lang API Guidelines on naming](https://rust-lang.github.io/api-guidelines/naming.html), but it mostly focuses on snake case `random_cut_forest` vs. kebab (?) case `random-cut-forest`. The guidelines have no official stance on the convention.

File: Rust/src/tree/bounding_box.rs
Comment (cswiercz) [reply Rust/src/tree/bounding_box.rs:224]: Thanks for the suggestion. Will try it out. One challenge is the multi-zipping.

File: Rust/src/tree/cut.rs
Comment (cswiercz) [reply Rust/src/tree/cut.rs:101]: Good callout. From what I understand I think it right approach is a `'static` lifetime variable. But then future parallelism might be an issue. I can continue (as I have in the rest of the code) under the assumption that the code is non-parallel for now and cross that bridge later.

I'll see what I can do about this.

File: Rust/src/tree/node.rs
Comment (cswiercz) [reply Rust/src/tree/node.rs:202]: Oh man. Good idea. But that's going to have big impact on trees, scoring, and the Python wrapper. Will need some time to think this through.

File: Rust/src/tree/bounding_box.rs
Comment (jotok) [reply Rust/src/tree/bounding_box.rs:224]: I was thinking that you could keep the range as the thing being iterated over:
```
(0..self.dimensions).all(|&i| /* point between min and max */)
```

File: Rust/src/tree/cut.rs
Comment (jotok) [reply Rust/src/tree/cut.rs:101]: I'm not familiar with how Rust implements this stuff, but given that they've separate the RNG from the distribution, it makes me wonder if the distribution is stateful. If it's not, then I think we shouldn't have to worry about concurrency.

File: Rust/src/tree/node.rs
Comment (jotok) [reply Rust/src/tree/node.rs:202]: If you like this idea but want to put it in a later PR, we could create an issue for it and add it to the project.

File: Rust/Cargo.toml
Comment (jotok) [reply Rust/Cargo.toml:2]: This isn't a blocker for me, but my feeling after perusing [here](https://crates.io/categories/data-structures?sort=downloads) is to go with either `random-cut-forest` or `randomcutforest`. Most people won't be familiar with the acronym, and personally when I look at a list of dependencies like to have at least a hint of what they provide.

Comment (cswiercz) [reply Rust/Cargo.toml:2]: Sure. I have no strong preference.

File: Rust/src/tree/node.rs
Comment (cswiercz) [reply Rust/src/tree/node.rs:202]: Create #170 to track this issue. Will defer until after the tree and other Rust code is shared and committed so that everyone involved can see the larger picture.

File: Rust/src/tree/cut.rs
Comment (cswiercz) [reply Rust/src/tree/cut.rs:101]: I did a deep dive into [`rand_distr::uniform::Uniform`](https://rust-random.github.io/rand/rand_distr/uniform/struct.Uniform.html) and I think this particular type is "cheap" enough that it shouldn't have a performance impact on this function. Furthermore, it does not implement any `const` or `static` methods so we cannot create one as such.

A `Uniform` type contains one field: a `Sampler` coming from something implementing the [`SampleUniform`](https://rust-random.github.io/rand/rand_distr/uniform/trait.SampleUniform.html) trait. (This trait has the `Sampler` as an associated type.) In the documentation for `SampleUniform` we see that the `Sampler` associated type has implementations for many different raw types, including `f64`. The implementation for `f64` is the [`UniformFloat`](https://rust-random.github.io/rand/rand_distr/uniform/struct.UniformFloat.html) type. Checking the [source code](https://rust-random.github.io/rand/src/rand/distributions/uniform.rs.html#809-812), this type seems to only designate a `low` field and a `scale` field, presumably indicating what is the smallest value the uniform sampler is allowed to attain and the size of the interval from that starting point. The cost of this is equivalent to if I did the following in the existing code:

```rust
let low = 0.0;
let scale = 1.0;
rng.sample_uniform(low, scale); // <--- this doesn't exist. need to provide a distribution
```

So from what I see, yes there is an instantiation happening with each call of `new_random_cut()` but the type only has two `f64` fields. Most of the state is contained in the `rng` argument passed into this function.

Comment (cswiercz) [reply Rust/src/tree/cut.rs:101]: FWIW, I wrote it this way because I couldn't find a "naked" "generate a random float between 0 and 1" function such as C's `rand()`.

Comment (jotok) [reply Rust/src/tree/cut.rs:101]: From a code organization perspective, it still makes more sense to me that the distribution would be a class constant, but I don't consider it a blocker.

**Review by jotok — APPROVED** [review]

**Pull request merged**
